lib/Dialect/DC/Transforms/Analysis.cpp

```diff
@@ -0,0 +1,283 @@
+    if (icmp.getPredicate() == circt::comb::ICmpPredicate::ule)
+      return DotNode{
+          .nodeType = "<=",
+          .incoming = valueToName(op.getOperands(), valuesMap, false),
+          .outgoing = valueToName(op.getResults(), valuesMap, false)};
+  }
+  return DotNode{.nodeType = "combNull",
+                 .incoming = valueToName(op.getOperands(), valuesMap, false),
+                 .outgoing = valueToName(op.getResults(), valuesMap, false)};
+}
+
+namespace {
+/// Emit the dot nodes
+struct DCDotPrintPass : public circt::dc::impl::DCDotPrintBase<DCDotPrintPass> {
+  void runOnOperation() override {
+
+    ModuleOp op = getOperation();
+
+    std::error_code ec;
+    llvm::raw_fd_ostream outfile("dc-out.dot", ec);
+    mlir::raw_indented_ostream os(outfile);
+
+    llvm::SmallVector<std::pair<mlir::Value, std::string>> valuesMap;
+    llvm::SmallVector<DotNode> nodes;
+
+    auto &moduleOps = op->getRegion(0).getBlocks();
+    // getBlock()->getOps<hw::HWModuleOp>();
+    for (auto &moduleOp : moduleOps) {
+      auto hwModuleOp = moduleOp.getOps<hw::HWModuleOp>();
+      for (auto hmo : hwModuleOp) {
+        for (auto &op : hmo->getRegion(0).getOps()) {
+          // either dc or comb operation
+          if (op.getDialect()->getNamespace() == "comb") {
+            // create new node
+            auto node = createCombNode(op, valuesMap);
+            nodes.push_back(node);
+          } else if (op.getDialect()->getNamespace() == "dc") {
+            auto node = createDCNode(op, valuesMap);
+            nodes.push_back(node);
+          }
+        }
+      }
+    }
+    os << "digraph{\n";
+    // print all nodes first
+    for (auto [i, n] : llvm::enumerate(nodes)) {
+      os << i << " [shape = polygon, label = \"" << n.nodeType << "\"]\n";
+    }
+    for (auto [id, n] : llvm::enumerate(nodes)) {
+      if (n.nodeType == "unpack")
+        for (auto ic : n.incoming)
+          os << "token_" << ic.second << " -> " << id << R"( [label = ")"
+             << ic.second << "\"]\n";
+    }
+    for (auto [id1, n1] : llvm::enumerate(nodes)) {
+      for (auto [id2, n2] : llvm::enumerate(nodes)) {
+        if (id1 != id2) {
+          for (const auto &n1Out : n1.outgoing) {
+            for (const auto &[i, n2In] : llvm::enumerate(n2.incoming)) {
+              if (n1Out.first == n2In.first) {
+                os << id1 << " -> " << id2 << " [label = \"" << n1Out.second
+                   << "\"]\n";
+              }
+            }
+          }
+        }
+      }
+    }
+    for (auto [id, n] : llvm::enumerate(nodes)) {
+      if (n.nodeType == "pack")
+        for (const auto &ic : n.outgoing)
+          os << id << " -> " << ic.second << " [label = \"" << ic.second
+             << "\"]\n";
+    }
+
+    os << "}";
+    outfile.close();
+  }
+};
+} // namespace
+
+std::unique_ptr<mlir::OperationPass<mlir::ModuleOp>>
+circt::dc::createDCDotPrintPass() {
+  return std::make_unique<DCDotPrintPass>();
+}
```

Comment:
nit: newline at EOF

Reply:
This was not resolved. You might have some auto format setting that reverts this change. IIRC, VS code has some odd behavior sometimes.